[2022/06/22] 백엔드 테스트 코드 추가(user / create / concat)

LeeSyong opened the pull request:
## 주제
백엔드 테스트 코드 추가

## 작업사항
- controller/auth, controller/video 수정
- mockDB / mockData 작성
- user.test.js / create.test.js / concat.test.js 추가

## 기타
우선 로그인, 동영상 생성 및 병합을 포함하는 테스트 코드 파일 3개를 작성했습니다. 추후 추가로 작성해야 합니다.

LeeSyong committed:
Test: add mockDB & mockData

LeeSyong committed:
Test: add user.test.js

LeeSyong committed:
Test: add create.test.js

LeeSyong committed:
Test: add concat.test.js

bellmin9321 commented on controllers/auth.js:
```diff
@@ -27,13 +27,7 @@ exports.createUser = async (req, res, next) => {
-  };
-
-  const { userId, profilePhoto } = user;
+  const { sub: userId, picture: profilePhoto } = decoded.payload;
```
옆의 코드를 한 줄로 줄인건가요!?

LeeSyong replied:
옆의 코드가 어떤 걸까요?? getPayload()를 이용하는 대신 과정을 좀 줄였습니다.

hansol-CHA replied:
객체 구조 분해 좋네요!!!

hansol-CHA commented on controllers/video.js:
```diff
@@ -45,6 +45,13 @@ exports.createVideo = async (req, res, next) => {
   const { video, thumbnail } = req.files;
   const { id } = req.decoded;
 
+  if (!title || !maxCreators) {
```
저도 리팩토링에서 이 부분을 고민하였는데, 몽구스 스키마에 정의한 내역에 따라 에러를 반환하더라구요! 그래도 명시적으로 추가하는게 나을까요?

LeeSyong replied:
넵 저도 고민한 부분이었는데 백엔드 과제 당시에 멘토님들께서 그래도 남기는 게 좋다고 말씀 주시더라구요. 프론트에서 input에 required 속성을 주어도 백엔드에서 다시 해당 부분 관련하여 에러 핸들링을 해주는 것과 비슷한 이치라고 설명해주셨던 기억이 납니다! 위 부분은 테스트 코드 작성을 위해 추가해보았습니다~

hansol-CHA commented on spec/concat.test.js:
```diff
@@ -0,0 +1,109 @@
+const request = require("supertest");
```
videoConcat.test가 나을까요..??

LeeSyong replied:
videoConcat.test.js가 나을까요 concatVideo.test.js가 나올까요? create.test.js 파일도 수정해주어야 할 것 같네요. 의견 부탁드립니다~

hansol-CHA replied:
흐으으으으으으으으음 명사형이 나은 것 같아 첫번째가 좋을 것 같아요

hansol-CHA commented on spec/concat.test.js:
```diff
@@ -0,0 +1,109 @@
+        expect(res.body.result).toBe("ng");
+        expect(res.body.errorMessage).toBe("Unauthorized");
+
+        const newVideo = await Video.findOne({ title: "newVideo" });
```
토큰이 유효하지 않으면 여기 쿼리 부분까지 안오지 않나요??? 그럼 없어도 될 것 같기도 하고... 테스트는 어렵네요 ㅠ

LeeSyong replied:
넵 토큰이 유효하지 않은 경우 쿼리 부분까지 안 오는 것이 맞습니다. 다만 쿼리 부분까지 오지 않는다는 것을 테스트하기 위해 추가한 코드입니다. 바로 아래에 toBeNull() 이란 메서드를 이용해 구현해보았습니다!

hansol-CHA replied:
이해했습니다!

hansol-CHA commented on spec/create.test.js:
```diff
@@ -0,0 +1,137 @@
+        .post("/api/videos")
+        .set("authorization", `Bearer ${mockToken}`)
+        .set("Content-Type", "multipart/form-data")
+        .field({ title: "newVideo" })
```
field와 attach은 어떤 차이가 있나요

LeeSyong replied:
원래는 req.body 값을 보낼 때는 send를 이용합니다만 attach를 사용하여 파일을 보내야 할 경우 attach와 send는 함께 사용할 수 없다고 하네요. 이 경우 send 대신 field를 사용해야 함께 테스트할 수 있다고 합니다~

hansol-CHA replied:
그렇군요 답변 감사합니다~

bellmin9321 replied:
좋은 정보네요!

LeeSyong committed:
Add: change file name

- create.test.js 삭제 => videoCreate.test.js 생성
- concat.test.js 삭제 => videoConcat.test.js 생성

LeeSyong committed:
Refactor: edit error messages

LeeSyong committed:
Chore: edit package.json

LeeSyong committed:
Chore: reinstall modules

LeeSyong committed:
Merge branch 'master' into test

hansol-CHA reviewed (approved):
고생 많으셨습니다~

bellmin9321 reviewed (approved):
테스트 확인 완료!
